### PR TITLE
219 confirmation on large tp

### LINKED ISSUE #219: Ask for confirmation to fit model if more than xxx time points (for non-spatial)
Background: Currently if you try to fit a model with say 800 time points your R session will crash because of insufficient memory. Typically non-spatial covariance structure model are used for less than say 100 time points only, so we could ask for user confirmation before trying to fit a model with more time points.

To do:
- [x] Find the right place in the code where to check this
- [x] Ask for user confirmation with simple question and expecting "yes" for confirmation
- [x] add test
- [x] add NEWS entry
- [x] add note of this to introduction vignette

### PR DESCRIPTION
**Pull request by clarkliming**
close #219 

`fit_mmrm` , `fit_mmrm_single` are updated to receive `tmb_data` and `formula_parts` to avoid repetative computations

`mmrm.max_visits` is used to determine the maximum number of visit levels allowed.

### CODE REVIEW

File: R/fit.R [hunk 1]
```diff
@@ -7,12 +7,16 @@
 #'
 #' @inheritParams mmrm
 #' @param control (`mmrm_control`)\cr object.
+#' @param tmb_data (`mmrm_tmb_data`)\cr object.
```
Comment (clarkliming) [inline R/fit.R:10]: this is added to prevent `h_mmrm_tmb_data` called multiple times in `mmrm`

File: R/fit.R [hunk 2]
```diff
@@ -32,27 +36,46 @@ fit_single_optimizer <- function(formula,
...
-      reml = reml,
-      covariance = covariance,
-      control = control
-    ),
-    remove = list(
-      warnings = c("NA/NaN function evaluation") # Transient visit to invalid parameters.
+  if (missing(tmb_data) || missing(formula_parts)) {
+    assert_formula(formula)
+    assert_data_frame(data)
+    assert_numeric(weights, any.missing = FALSE, lower = .Machine$double.xmin)
```
Comment (clarkliming) [inline R/fit.R:46]: check weights as positive values

File: R/fit.R [hunk 3]
```diff
@@ -32,27 +36,46 @@ fit_single_optimizer <- function(formula,
...
+    assert_flag(reml)
+    assert_class(control, "mmrm_control")
+    assert_list(control$optimizers, names = "unique", types = c("function", "partial"))
+    quiet_fit <- h_record_all_output(
+      fit_mmrm(
```
Comment (clarkliming) [inline R/fit.R:51]: `fit_mmrm` is updated to accept two types of argument, and `fit_mmrm` still works as low level functionality

File: R/fit.R [hunk 4]
```diff
@@ -145,9 +168,8 @@ refit_multiple_optimizers <- function(fit,
     FUN = fit_single_optimizer,
     control = controls,
     MoreArgs = list(
-      formula = old_formula,
-      data = old_data,
-      weights = old_weights,
+      tmb_data = fit$tmb_data,
+      formula_parts = fit$formula_parts,
       reml = fit$reml
```
Comment (clarkliming) [inline R/fit.R:173]: reml can be removed as tmb_data already has it

File: R/fit.R [hunk 5]
```diff
@@ -403,6 +427,8 @@ mmrm <- function(formula,
   if (!is.null(fit_msg)) {
     message(paste(fit_msg, collapse = "\n"))
   }
+  fit$call <- match.call()
+  fit$call$formula <- formula
```
Comment (clarkliming) [inline R/fit.R:431]: obtain the formula so that it could be printed, instead of a object.

File: R/mmrm-methods.R [hunk 6]
```diff
@@ -122,16 +122,16 @@ h_print_call <- function(call, n_obs, n_subjects, n_timepoints) {
...
       n_subjects, "subjects with maximum", n_timepoints, "timepoints)",
       fill = TRUE
     )
   }
-  # Only if call$weights is a string then it was explicitly given by the user.
-  if (test_string(tmp <- call$weights)) {
-    cat("Weights:    ", tmp, fill = TRUE)
+  # Display the expression of weights
+  if (!is.null(call$weights)) {
+    cat("Weights:    ", deparse(call$weights), fill = TRUE)
```
Comment (clarkliming) [inline R/mmrm-methods.R:134]: weights can be something like `rep(1, nrow(data)` so have to deparse it

File: R/tmb.R [hunk 7]
```diff
@@ -514,20 +524,9 @@ fit_mmrm <- function(formula,
...
-    attr(data, which = "dataname"),
-    toString(match.call()$data)
-  )
-  weights_str <- attr(weights, which = "dataname")
-  if (!is.null(weights_str)) {
-    fun_call$weights <- weights_str
-  }
-  fit$call <- fun_call
+  fit$call <- match.call()
+  fit$call$formula <- formula_parts$formula
```
Comment (clarkliming) [inline R/tmb.R:530]: for the data, use the symbol directly; in `mmrm` the call is also retrieved. The reason to have two locations of call is that `fit_mmrm` is still exported so we need to use it.  For weights, I am just wondering if a "string" indicating the column in the data should be better than providing a numeric value

**Comment by clarkliming** [comment]
how to trigger the actions manualy? in https://github.com/openpharma/mmrm/actions/runs/4292615230 all green

**Comment by danielinteractive** [comment]
@clarkliming just another manual (also empty is ok) commit that is pushed on this branch will retrigger checks

File: R/fit.R (hunk 2 above)
Comment (danielinteractive) [reply R/fit.R:46]: nice

File: R/fit.R (hunk 3 above)
Comment (danielinteractive) [reply R/fit.R:51]: hm but it would be better if we keep named arguments in this `fit_mmrm` call?

File: R/fit.R [hunk 8]
```diff
@@ -32,27 +36,46 @@ fit_single_optimizer <- function(formula,
...
+        control = control
+      ),
+      remove = list(
+        warnings = c("NA/NaN function evaluation") # Transient visit to invalid parameters.
+      )
     )
-  )
+  } else {
+    checkmate::assert_class(tmb_data, "mmrm_tmb_data")
+    checkmate::assert_class(formula_parts, "mmrm_tmb_formula_parts")
```
Comment (danielinteractive) [inline R/fit.R:65]:
```suggestion
    assert_class(tmb_data, "mmrm_tmb_data")
    assert_class(formula_parts, "mmrm_tmb_formula_parts")
```

File: R/fit.R (hunk 4 above)
Comment (danielinteractive) [reply R/fit.R:173]: but now just `weights` is removed?

File: R/fit.R [hunk 9]
```diff
@@ -368,13 +390,15 @@ mmrm <- function(formula,
   } else {
     attr(weights, which = "dataname") <- deparse(match.call()$weights)
   }
-
+  covariance <- reconcile_cov_struct(formula, covariance)
```
Comment (danielinteractive) [inline R/fit.R:393]: is that a helper function call? then better prefix with `h_`

File: R/fit.R (hunk 5 above)
Comment (danielinteractive) [reply R/fit.R:431]: ok

File: R/mmrm-methods.R (hunk 6 above)
Comment (danielinteractive) [reply R/mmrm-methods.R:134]: yeah this was a bit hacky, so maybe this is a bit better

File: R/tmb.R [hunk 10]
```diff
@@ -468,22 +471,29 @@ fit_mmrm <- function(formula,
...
+    assert_list(control$optimizers, min.len = 1)
+    assert_numeric(weights, any.missing = FALSE)
+    assert_true(all(weights > 0))
+    tmb_data <- h_mmrm_tmb_data(
+      formula_parts, data, weights, reml,
+      accept_singular = control$accept_singular, drop_visit_levels = control$drop_visit_levels
+    )
+  } else {
+    checkmate::assert_class(tmb_data, "mmrm_tmb_data")
+    checkmate::assert_class(formula_parts, "mmrm_tmb_formula_parts")
```
Comment (danielinteractive) [inline R/tmb.R:495]:
```suggestion
    assert_class(tmb_data, "mmrm_tmb_data")
    assert_class(formula_parts, "mmrm_tmb_formula_parts")
```

File: R/tmb.R (hunk 7 above)
Comment (danielinteractive) [reply R/tmb.R:530]: hm interesting. I remember we had some problem with the data thing before. but I guess as long as the tests pass this should be fine.

and then the same logic as for data applies to the weights thing?

File: R/utils.R [hunk 11]
```diff
@@ -296,3 +296,33 @@ fill_names <- function(x) {
...
+#' Ask for Confirmation on Large Visit Levels
+#'
+#' @description Ask the user for confirmation if there are too many visit levels
+#' for non-spatial covariance structure in interactive sessions.
+#'
+#' @param x (`numeric`)\cr number of visit levels.
+#'
+#' @keywords internal
+h_confirm_large_levels <- function(x) {
+  checkmate::assert_numeric(x)
```
Comment (danielinteractive) [inline R/utils.R:309]:
```suggestion
  assert_count(x)
```

File: tests/testthat/_snaps/tmb-methods.md [hunk 12]
```diff
@@ -5,6 +5,7 @@
     Formula:     FEV1 ~ RACE + us(AVISIT | USUBJID)
     Data:        fev_data (used 537 observations from 197 subjects with maximum 4 
     timepoints)
+    Weights:     rep(1, nrow(fev_data))
```
Comment (danielinteractive) [inline tests/testthat/_snaps/tmb-methods.md:8]: hm this is not very nice. I think therefore I had some additional logic before to circumvent this from being printed

File: vignettes/introduction.Rmd [hunk 13]
```diff
@@ -273,6 +273,16 @@ and `VIS4` the same as the correlation between `VIS1` and `VIS2`. Hence we get a
...
 
+## Large Number of Visit Levels for non-spatial Covariance
+
+When the number of visit levels is large, it usually require large memory to create the
+covariance matrix. By default, the maximum allowed visit levels is 100, and if there are more
+visit levels, a confirmation is needed if run interactively.
+You can use `options(mmrm.max_visits = <target>)` to increase the maximum allowed number of visit
+levels.
+In non-interactive sessions the confirmation is not raised and will directly give you error if
+that number of visit levels exceed the maximum.
```
Comment (danielinteractive) [inline vignettes/introduction.Rmd:284]: hm I think this is maybe a bit too detailed for the introduction vignette. Maybe just move this to the `mmrm()` function documentation?

**Review by danielinteractive — COMMENTED** [review]
thanks a lot Liming! please see few comments inside

File: tests/testthat/_snaps/tmb-methods.md (hunk 12 above)
Comment (clarkliming) [reply tests/testthat/_snaps/tmb-methods.md:8]: I think previously only when `weights` is assigned (not rep(1, nrow(data)) then it is displayed. do you mean this https://github.com/openpharma/mmrm/blob/b3f7e5c7d53ee605e4b2041aba48a3f9c42b7878/R/fit.R#L369

so in previous logic, weights get displayed when:

in `mmrm`, weights argument is used, not NULL; in `fit_mmrm`, no matter if weights is assigned, it is not displayed

here the change is that in `fit_mmrm` the assigned `weights` still get displayed. we can of course remove that "weights" for `fit_mmrm$call`. but why we should do that? in `mmrm` by default the weights are not displayed

File: R/fit.R (hunk 4 above)
Comment (clarkliming) [reply R/fit.R:173]: yes; `reml` is already part of `tmb_data`

File: tests/testthat/_snaps/tmb-methods.md (hunk 12 above)
Comment (danielinteractive) [reply tests/testthat/_snaps/tmb-methods.md:8]: ah ok cool - so when we use `mmrm()` without explicit weights assigned it is not displayed still right? then it is ok.

File: R/fit.R [hunk 14]
```diff
@@ -345,6 +344,13 @@ mmrm_control <- function(n_cores = 1L,
 #'
 #' NA values are always omitted regardless of `na.action` setting.
 #'
+#' When the number of visit levels is large, it usually require large memory to create the
+#' covariance matrix. By default, the maximum allowed visit levels is 100, and if there are more
+#' visit levels, a confirmation is needed if run interactively.
+#' You can use `options(mmrm.max_visits = <target>)` to increase the maximum allowed number of visit
+#' levels. In non-interactive sessions the confirmation is not raised and will directly give you error if
+#' that number of visit levels exceed the maximum.
```
Comment (danielinteractive) [inline R/fit.R:352]:
```suggestion
#' When the number of visit levels is large, it usually requires large memory to create the
#' covariance matrix. By default, the maximum allowed visit levels is 100, and if there are more
#' visit levels, a confirmation is needed if run interactively.
#' You can use `options(mmrm.max_visits = <target>)` to increase the maximum allowed number of visit
#' levels. In non-interactive sessions the confirmation is not raised and will directly give you an error if
#' the number of visit levels exceeds the maximum.
```

**Review by danielinteractive — APPROVED** [review]
thanks Liming!

File: tests/testthat/_snaps/tmb-methods.md (hunk 12 above)
Comment (clarkliming) [reply tests/testthat/_snaps/tmb-methods.md:8]: it only gets printed when you have "weights" argument specified; e.g. using `mmrm(formula, data = fev_data)` you don't have the weights printed. If you have `mmrm(weights = xxx)` it got printed (this is the same for previous version)

Comment (clarkliming) [reply tests/testthat/_snaps/tmb-methods.md:8]: (and that's why `fit_mmrm` and `mmrm` both keep a `call` but they always use the call they received)

**Review by danielinteractive — APPROVED** [review]

**Pull request merged**